test/sys/test_socket.rs

```diff
@@ -186,6 +186,57 @@ pub fn test_scm_rights() {
     close(w).unwrap();
 }
 
+#[cfg(any(target_os = "linux", target_os= "android"))]
+#[test]
+pub fn test_af_alg_cmsg() {
+    use std::{mem, ptr};
+    use nix::sys::uio::IoVec;
+    use nix::unistd::read;
+    use nix::sys::socket::{socket, sendmsg, bind, accept, setsockopt,
+                           AddressFamily, SockType, SockFlag, SockAddr,
+                           ControlMessage, MsgFlags};
+    use nix::sys::socket::sockopt::AlgSetKey;
+
+    let alg_type = "skcipher";
+    let alg_name = "ctr(aes)";
+
+    let sock = socket(AddressFamily::Alg, SockType::SeqPacket, SockFlag::empty(), None)
```

Comment:
Could you condense all of the socket setup code into a single `socketpair` call?

Reply:
Could you point me to some code examples of how it may be implemented?

Reply:
Look at test_scm_rights in test/sys/test_socket.rs.

Reply:
From what I see in [socketpair docs](http://man7.org/linux/man-pages/man2/socketpair.2.html), this is a call which "creates a pair of connected sockets". I didn't find any mentions in AF_ALG documentation regarding it. Even [libkcapi](https://github.com/smuellerDD/libkcapi), which is a reference implementation of AF_ALG doesn't use this call. I have no idea how to utilize it.

Reply:
As of the test_scm_rights example, it looks like it tests the serialization and deserialization of cmsg by sending it and receiving through `socketpair`. I don't think that it may somehow improve this particular test case. We don't receive this message from the kernel in the real scenario. It's used only to request crypto operation from the kernel, and the further assertions validate the correctness of these messages. 
What I think may improve these tests, is to generate input data randomly, and validate the AF_ALG results with the results from another crypto implementation. But it will require an additional dev dependency.

Reply:
You're right that socketpair wouldn't help; I didn't read the test closely enough the first time.  But we shouldn't try to validate that Linux's crypto is correct; we only need to verify that the bindings work.  So there's no need to test with random data and verify the cryptography.